extension/ts/background-scripts/background.js/telemetry/Telemetry.ts

```diff
@@ -0,0 +1,44 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Glean from "@mozilla/glean/webext";
+import { custom } from "./generated/pings";
+import { toLang, fromLang } from "./generated/metadata";
+
+export class Telemetry {
+  private static _instance = null;
+
+  constructor() {
+    Glean.initialize("bergamot-extension", true, { debug: { logPings: true } });
+  }
+
+  public static global() {
+    if (Telemetry._instance == null) {
+      Telemetry._instance = new Telemetry();
+    }
+    return Telemetry._instance;
+  }
+
+  public setUploadEnabled(val: boolean) {
+    Glean.setUploadEnabled(val);
+  }
+
+  public record = (
```

Comment:
As I understand it, you are submitting a ping for each metric that is collected. That will make analysis of data harder and may cause issues due do the big amount of telemetry calls in a small timeframe. Ideally you want a group of related metrics in a single ping, not one metric per ping.

I would advise you to re-assess whether or not it is necessary to include the to and from languages with _every_ metric recorded. Note that for event metrics you can include this data in the [event extras key](https://mozilla.github.io/glean/book/user/metrics/event.html#configuration).

If events are not enough, then the other option is to try the process of changing metric types in Glean, where can discuss how to better attend to this demand. We have a specific process for that which can be checked out [here](https://wiki.mozilla.org/Glean/Adding_or_changing_Glean_metric_types#The_proposal_process).

Reply:
One additional option could be to send pings whenever the language pair changes (which I assume is far less frequent than setting individual metrics).  That would mean that each ping only applies to one language pair context (which I think it what you want).

Reply:
Yes, this would be an ideal option. From analysis perspective we need to have a separate columns for To and From languages. Parsing it from events for some metrics is expensive and not convenient. When we start measuring timespan and quantity, they also should be sent separately, not as aggregates. I would like to delegate all the complexity of sending and grouping metrics to telemetry library. I assume when other types of pings will be added it will be the way to group things together.

It would be great to set metric metadata in the same way as client info, not as a separate metric. Then it would be possible to distinguish which metrics can be grouped and sent in one ping. 

Having of language pair for all the metrics is a product requirement. We use different ML models for different language pairs and we'll need to investigate problems and perform analysis per language pair. 

Telemetry calls should be super cheap and don't cause any performance degradation in client code.

So, for now the only option we have is sending a custom ping per metric.

Reply:
> When we start measuring timespan and quantity, they also should be sent separately, not as aggregates.

Would you mind explaining a bit further why this is a necessity? Maybe an analysis example or something more concrete? I would like to have a better grasp on your demands so I can help on finding an ideal solution.

> I would like to delegate all the complexity of sending and grouping metrics to telemetry library. I assume when other types of pings will be added it will be the way to group things together.

Note that it may not be possible to delegate all of it to Glean.js. From what I am seeing you have very specific data aggregation and ping scheduling needs, whatever we provide out of the box will be generic and might not address your use cases.

> It would be great to set metric metadata in the same way as client info, not as a separate metric. Then it would be possible to distinguish which metrics can be grouped and sent in one ping.

Unfortunately it is not currently possible to do something like this. We have recently filed [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1695236) to discuss allowing an option to send specific metrics in all pings. I am not sure this would address your use case, but if so we will take it in account when discussing this new feature.

> Telemetry calls should be super cheap and don't cause any performance degradation in client code.

The client side calls may be cheap, but every ping submission results in a HTTP request, sending so many pings will end up using a lot of the users bandwidth. Keep in mind, that Glean is not a tool for streaming telemetry which is basically what you implemented here, having a ping per metric will make it extremely hard to correlate metrics amongst these different pings. Also, although Glean.js does not have rate limiting in place at the moment, it will be implemented at some point and by sending so many pings you will quickly reach these limits. 

I am available if you would like to discuss these design choices synchronously in a meeting or what not :)

Reply:
> 
> 
> Yes, this would be an ideal option. From analysis perspective we need to have a separate columns for To and From languages. Parsing it from events for some metrics is expensive and not convenient. When we start measuring timespan and quantity, they also should be sent separately, not as aggregates. I would like to delegate all the complexity of sending and grouping metrics to telemetry library. I assume when other types of pings will be added it will be the way to group things together.

I believe we should focus more on the questions that need to be answered, as we flagged during our last meeting, more than on the shape data should look like. Would you kindly share with us which questions you're planning on answering with the metrics being introduced here?

> Having of language pair for all the metrics is a product requirement. We use different ML models for different language pairs and we'll need to investigate problems and perform analysis per language pair.
> 
> Telemetry calls should be super cheap and don't cause any performance degradation in client code.

Did you see a specific regression after adding Glean.js? If so, would you kindly share the performance profile with us?
Moreover, data collection is meant to be "cheap but not free". Collecting on hotpaths is **always** a bad idea, no matter what data collection library is being used.

> So, for now the only option we have is sending a custom ping per metric.

This is bad, as @brizental highlighted, and must be reconsidered.

Reply:
I think I have an idea of what we can do. We can try to introduce a concept of translation session - from the moment the info bar appears on a new page to the moment it is closed or language pair changed from the UI. Then we can group all the UI events, performance metrics and errors and send them once when this session ends. The only exception would be when a user changes language, then we need to send everything and start a new session. 

As for "language mismatch" and "language pair unsupported" counters, I would prefer to send them automatically in metrics ping, but it's not implemented in glean.js yet. 

What do you think?

Reply:
> Would you kindly share with us which questions you're planning on answering with the metrics being introduced here?
You can find questions we want to answer in the Bergamot Telemetry doc I sent to everyone at the meeting

Reply:
> I think I have an idea of what we can do. We can try to introduce a concept of translation session - from the moment the info bar appears on a new page to the moment it is closed or language pair changed from the UI. Then we can group all the UI events, performance metrics and errors and send them once when this session ends. The only exception would be when a user changes language, then we need to send everything and start a new session.

This sounds like a great idea. Thank you for reconsidering :)

> As for "language mismatch" and "language pair unsupported" counters, I would prefer to send them automatically in metrics ping, but it's not implemented in glean.js yet.

Yes, that feature will be discussed / implemented in our next milestone.

Reply:
I realized that for "language mismatch" and "language pair unsupported" we still need language pair metadata, so we can't send them in `metrics` ping as counters. Only as events with extra in `events` ping. It's not clear for me at what moment to send a custom ping with those metrics, because this is a regular user workflow, without the infobar. Let's leave them out of scope for the first experimental integration, so we'll think about this later.

Reply:
ok, we don't submit a ping for every metric now. It is submitted when  infobar is closed, language pair is changed or translation is completed (we need this because it is not clear what other event to use if infobar was not closed explicitly). This is good enough for the first iteration, but there is an issue with multiple tabs opened in different languages. It seems they use shared chrome context and since glean and telemetry are global, it doesn't work as expected. We can address this in a separate issue. @motin